### PR TITLE
EVA-3459 - New script and tests to check if sequences in fasta are from INSDC

### PR DESCRIPTION
**Pull request by tcezard**

### CODE REVIEW

File: eva_sub_cli/nextflow/validation.nf [hunk 1]
```diff
@@ -187,3 +195,22 @@ process sample_name_concordance {
...
+process genome_checker {
+    publishDir "$params.output_dir",
+            overwrite: true,
+            mode: "copy"
+
+    input:
+    path(fasta_file)
+
+    output:
+    path "${fasta_file}_check.yml", emit: fasta_checker
```
Comment (apriltuesday) [inline eva_sub_cli/nextflow/validation.nf:210]:
```suggestion
    path "${fasta_file}_check.yml", emit: fasta_checker_yml
```

File: eva_sub_cli/nextflow/validation.nf [hunk 2]
```diff
@@ -25,11 +25,12 @@ params.executable = [
     "vcf_validator": "vcf_validator",
     "vcf_assembly_checker": "vcf_assembly_checker",
     "samples_checker": "samples_checker.py",
+    "fasta_checker": "check_fasta_insdc.py",
     "xlsx2json": "xlsx2json.py",
     "biovalidator": "biovalidator"
 ]
 // validation tasks
-params.validation_tasks = [ "vcf_check", "assembly_check", "samples_check", "metadata_check"]
+params.validation_tasks = [ "vcf_check", "assembly_check", "samples_check", "metadata_check", "genome_check"]
```
Comment (apriltuesday) [inline eva_sub_cli/nextflow/validation.nf:33]: Maybe `fasta_check` to match the name of the script, or `insdc_check`? Something to distinguish it more from `assembly_check`

File: bin/check_fasta_insdc.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+def assess_fasta(input_fasta):
+    results = {'sequences': []}
+    all_insdc = True
+    for header, sequence in fasta_iter(input_fasta):
+        name = header.split()[0]
+        md5_digest = refget_md5_digest(sequence)
+        sequence_metadata = get_refget_metadata(md5_digest)
+        results['sequences'].append({'sequence_name': name, 'sequence_md5': md5_digest, 'insdc': bool(sequence_metadata)})
+        all_insdc = all_insdc and bool(sequence_metadata)
+    results['all_insdc']: all_insdc
```
Comment (apriltuesday) [inline bin/check_fasta_insdc.py:77]:
```suggestion
    results['all_insdc'] = all_insdc
```

File: bin/check_fasta_insdc.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    return results
+
+
+def main():
+    arg_parser = argparse.ArgumentParser(
+        description='Calculate the each sequence Refget MD5 identify and compare these against INSDC Refget server.')
```
Comment (apriltuesday) [inline bin/check_fasta_insdc.py:83]:
```suggestion
        description='Calculate each sequence's Refget MD5 digest and compare these against INSDC Refget server.')
```

File: bin/check_fasta_insdc.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+def refget_md5_digest(sequence):
+    return hashlib.md5(sequence.upper().encode('utf-8')).hexdigest()
+
+
+def fasta_iter(input_fasta):
+    """
+    Given a fasta file. yield tuples of header, sequence
+    """
+    "first open the file outside "
+    fin = open(input_fasta, 'r')
```
Comment (apriltuesday) [inline bin/check_fasta_insdc.py:42]: I believe it's okay to `yield` from within a `with` block if that's the concern here. Either way we should close the file at the end of the iterator...

File: bin/check_fasta_insdc.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+    # ditch the boolean (x[0]) and just keep the header or sequence since
+    # we know they alternate.
+
```
Comment (apriltuesday) [inline bin/check_fasta_insdc.py:46]:
```suggestion
```

File: bin/check_fasta_insdc.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    faiter = (x[1] for x in groupby(fin, lambda line: line[0] == ">"))
+
+    for header in faiter:
+        # drop the ">"
+        headerStr = header.__next__()[1:].strip()
+
+        # join all sequence lines to one.
+        seq = "".join(s.strip() for s in faiter.__next__())
+        yield (headerStr, seq)
```
Comment (apriltuesday) [inline bin/check_fasta_insdc.py:55]: This is clever, though I must confess I read it through a number of times in utter confusion before checking the docs to see that `groupby` only groups consecutive elements

File: tests/test_report.py [hunk 8]
```diff
@@ -67,6 +67,15 @@
             }
         }
     },
+    "fasta_check": {
+        'reference_genome.fa': {
```
[the remark this answers is not in the document]
Comment (tcezard) [reply tests/test_report.py:71]: Done

File: bin/check_fasta_insdc.py (hunk 5 above)
Comment (tcezard) [reply bin/check_fasta_insdc.py]: Added

File: bin/check_fasta_insdc.py (hunk 7 above)
Comment (tcezard) [reply bin/check_fasta_insdc.py]: Yes it was clever indeed: https://www.biostars.org/p/710/#120760

**Review by apriltuesday — APPROVED** [review]

**Review by nitin-ebi — APPROVED** [review]

**Pull request merged**